worlds/fe8/fe8py.py

```diff
@@ -434,20 +434,34 @@ def __init__(self, rom: bytearray, random: Random):
         self.weapons_by_rank[WeaponRank.E].append(self.weapons_by_name["Flux"])
 
     def job_valid(self, job: JobData, char: int, logic: dict[str, Any]) -> bool:
+        #get character tags
         tags = self.character_store.tags(char)
-        if tags is not None and "generic" in tags:
-            if "generic_only" in job.tags:
-                return False
-
-        if "must_fly" in logic and logic["must_fly"] and "flying" not in job.tags:
+        if not tags: tags = set()
+
+        #get list of tags that make the job invalid (notags)
+        #the "no_" prefix adds the tag to the invalid tag list
+        #"no_flying" makes any job with "flying" tag invalid
+        notags = set(z.removeprefix("no_") for z in tags if z.startswith("no_"))
+        for x in logic:
+            if x.startswith("no_") and logic[x]:
+                notags.add(x.removeprefix("no_"))
+        #job is invalid if it has any of the tags in notags
+        if notags and notags&job.tags:
             return False
 
-        if "no_fly" in logic and logic["no_fly"] and "flying" in job.tags:
+        if (
+            ("must_fly" in logic 
```

Comment:
Can we just add the tags into the logic?

Reply:
that is an excellent idea, we can just add the tags when building the unit specific logic

Reply:
a44be9adafc7cc6b6f8994fc779aab479ab4f3e1 changed it so that character tags get added into logic near the start of **randomize_chapter_unit**
Also, i believe this means that character tags should now affect the inventory generation, which might be useful later on